frame/support/src/storage/transactional.rs

```diff
@@ -158,6 +158,40 @@ pub fn with_transaction_unchecked<R>(f: impl FnOnce() -> TransactionOutcome<R>)
 	}
 }
 
+/// Execute the supplied function, adding a new storage layer.
+///
+/// This is the same as `with_transaction`, but assuming that any function returning
+/// an `Err` should rollback, and any function returning `Ok` should commit. This
+/// provides a cleaner API to the developer who wants this behavior.
+pub fn with_storage_layer<T, E>(f: impl FnOnce() -> Result<T, E>) -> Result<T, E>
```

Comment:
Cool. Another method that can graduate from ORML https://github.com/open-web3-stack/open-runtime-module-library/blob/954cc7d97f70fe5c3c5ba286e0160b48ab5c6758/utilities/src/lib.rs#L28